src/ed25519.rs

```diff
@@ -777,22 +777,37 @@ impl PublicKey {
     /// Derive this public key from its corresponding `SecretKey`.
     #[allow(unused_assignments)]
     pub fn from_secret<D>(secret_key: &SecretKey) -> PublicKey
-            where D: Digest<OutputSize = U64> + Default {
-
+        where D: Digest<OutputSize = U64> + Default
+    {
         let mut h:    D = D::default();
         let mut hash:   [u8; 64] = [0u8; 64];
         let mut digest: [u8; 32] = [0u8; 32];
-        let     pk:     [u8; 32];
 
         h.input(secret_key.as_bytes());
         hash.copy_from_slice(h.fixed_result().as_slice());
 
         digest.copy_from_slice(&hash[..32]);
-        digest[0]  &= 248;
-        digest[31] &= 127;
-        digest[31] |= 64;
 
-        pk = (&Scalar::from_bits(digest) * &constants::ED25519_BASEPOINT_TABLE).compress().to_bytes();
+        PublicKey::mangle_scalar_bits_and_multiply_by_basepoint_to_produce_public_key(&mut digest)
+    }
+
+    /// Derive this public key from its corresponding `ExpandedSecretKey`.
+    pub fn from_expanded_secret(expanded_secret_key: &ExpandedSecretKey) -> PublicKey {
```

Comment:
Why have both this function and a `From` impl? It seems like it would be simpler just to have this code in the `From` impl and have people call `PublicKey::from(expanded_secret_key)`.